data/transactions/logic/eval.go

```diff
@@ -4714,8 +4720,13 @@ func (cx *EvalContext) appReference(ref uint64, foreign bool) (aid basics.AppInd
 	if foreign {
 		// In old versions, a foreign reference must be an index in ForeignApps or 0
 		if ref <= uint64(len(cx.txn.Txn.ForeignApps)) {
-			return basics.AppIndex(cx.txn.Txn.ForeignApps[ref-1]), nil
+			return cx.txn.Txn.ForeignApps[ref-1], nil
+		}
+		// it seems most consistent to allow slot access into tx.Access if it is used with an old app.
```

Comment:
this looks inconsistent - we do not allow both ForeignApps and Access use in a transaction but here (and in `resolveApp` below) both are checked

Reply:
I could not decide which makes for clearer code.  Only one of A or B can be true, so is it nice to write:
```
if isA {
  A
} else {
 B
}
```

or nice to just write
```
A
B
```

I decided to go with the later to save space. Either A or B becomes a no-op because the length of one of the arrays is 0.

Reply:
Since we are not allowing tx.Access with old programs, I can remove all the new code, so it doesn't matter.